File: userland/indiastack/sigma_india_stack.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 SigmaOS Project
//
// userland/indiastack/sigma_india_stack.rs — India Stack Integration
// ABDM FHIR client, UPI/NPCI, GST/IRN, Aadhaar APIs
// Language: Rust (std)

use std::collections::HashMap;

// ── ABDM FHIR ─────────────────────────────────────────────────────────────
#[derive(Debug, Clone)]
pub struct FhirPatient {
    pub health_id:   String,
    pub name:        String,
    pub dob:         String,
    pub gender:      String,
    pub mobile:      String,
    pub abha_address: String,
}

#[derive(Debug, Clone)]
pub struct FhirObservation {
    pub patient_id:  String,
    pub code:        String,   // LOINC/SNOMED code
    pub value:       String,
    pub unit:        String,
    pub timestamp:   String,
}

pub struct AbdmClient {
    pub base_url:    String,
    pub client_id:   String,
    pub access_token: Option<String>,
}

impl AbdmClient {
    pub fn new(client_id: &str) -> Self {
        Self {
            base_url:    "https://dev.abdm.gov.in/gateway".to_owned(),
            client_id:   client_id.to_owned(),
            access_token: None,
        }
    }

    pub fn sandbox() -> Self {
        let mut c = Self::new("sigma-os-sandbox");
        c.base_url = "https://dev.abdm.gov.in/gateway".to_owned();
        c
    }

    /// OAuth2 token request
    pub fn authenticate(&mut self, client_secret: &str) -> bool {
        let body = format!(
            r#"{{"clientId":"{}","clientSecret":"{}","grantType":"client_credentials"}}"#,
            self.client_id, client_secret
        );
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",
                   &format!("{}/v0.5/sessions", self.base_url),
                   "-H","Content-Type: application/json",
                   "-d",&body,"--max-time","10"])
            .output();
        match out {
            Ok(o) if o.status.success() => {
                let resp = String::from_utf8_lossy(&o.stdout);
                if let Some(pos) = resp.find("\"accessToken\":\"") {
                    let rest = &resp[pos+15..];
                    if let Some(end) = rest.find('"') {
                        self.access_token = Some(rest[..end].to_owned());
                        return true;
                    }
                }
                false
            }
            _ => false,
        }
    }

    fn auth_header(&self) -> String {
        self.access_token.as_ref().map(|t| format!("Bearer {}", t))
            .unwrap_or_default()
    }

    /// Search for a patient by ABHA (Ayushman Bharat Health Account) ID
    pub fn search_patient(&self, abha_id: &str) -> Option<FhirPatient> {
        let token = self.auth_header();
        if token.is_empty() { return None; }
        let url = format!("{}/v0.5/patients/find", self.base_url);
        let body = format!(r#"{{"query":{{"id":"{}","purpose":"KYC"}}}}"#, abha_id);
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",&url,
                   "-H","Content-Type: application/json",
                   "-H",&format!("Authorization: {}", token),
                   "-d",&body,"--max-time","10"])
            .output().ok()?;
        if !out.status.success() { return None; }
        let resp = String::from_utf8_lossy(&out.stdout);
        // Parse minimal FHIR Patient resource
        Some(FhirPatient {
            health_id:    abha_id.to_owned(),
            name:         extract_json(&resp, "name").unwrap_or_default(),
            dob:          extract_json(&resp, "dob").unwrap_or_default(),
            gender:       extract_json(&resp, "gender").unwrap_or_default(),
            mobile:       extract_json(&resp, "mobile").unwrap_or_default(),
            abha_address: extract_json(&resp, "healthId").unwrap_or_default(),
        })
    }

    /// Link health records to ABHA ID
    pub fn link_records(&self, abha_id: &str, care_context_reference: &str) -> bool {
        let token = self.auth_header();
        if token.is_empty() { return false; }
        let body = format!(r#"{{"requestId":"sigma-{}","patient":{{"id":"{}","referenceNumber":"{}"}}}}"#,
                           std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH)
                               .unwrap_or_default().as_secs(),
                           abha_id, care_context_reference);
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",
                   &format!("{}/v0.5/links/link/init", self.base_url),
                   "-H","Content-Type: application/json",
                   "-H",&format!("Authorization: {}", token),
                   "-d",&body,"--max-time","10"])
            .output();
        matches!(out, Ok(o) if o.status.success())
    }
}

// ── UPI / NPCI ────────────────────────────────────────────────────────────
#[derive(Debug, Clone)]
pub struct UpiTransaction {
    pub txn_id:     String,
    pub payer_vpa:  String,   // Virtual Payment Address e.g. user@okicici
    pub payee_vpa:  String,
    pub amount_paise: u64,    // Amount in paise (1 INR = 100 paise)
    pub note:       String,
    pub ref_id:     String,
}

#[derive(Debug, Clone)]
pub enum UpiStatus { Pending, Success, Failed, Reversed }

pub struct UpiClient {
    pub merchant_id: String,
    pub vpa:         String,   // Merchant VPA
    pub upi_url:     String,   // NPCI production: https://api.npci.org.in/upiip/v1
}

impl UpiClient {
    pub fn new(merchant_id: &str, vpa: &str) -> Self {
        Self {
            merchant_id: merchant_id.to_owned(),
            vpa:         vpa.to_owned(),
            upi_url:     "https://api.npci.org.in/upiip/v1".to_owned(),
        }
    }

    /// Generate a UPI payment URI (for QR code)
    pub fn payment_uri(&self, amount_paise: u64, note: &str, txn_ref: &str) -> String {
        format!("upi://pay?pa={}&pn=SigmaOS&am={:.2}&tn={}&tr={}&cu=INR",
                self.vpa,
                amount_paise as f64 / 100.0,
                urlencoded(note),
                txn_ref)
    }

    /// Initiate a collect request (pull payment)
    pub fn collect_request(&self, payer_vpa: &str, amount_paise: u64, note: &str) -> Option<String> {
        let txn = UpiTransaction {
            txn_id:       format!("sigma-{}", timestamp_millis()),
            payer_vpa:    payer_vpa.to_owned(),
            payee_vpa:    self.vpa.clone(),
            amount_paise,
            note:         note.to_owned(),
            ref_id:       format!("ref{}", timestamp_millis()),
        };
        let body = format!(
            r#"{{"txnId":"{}","payerVpa":"{}","payeeVpa":"{}","amount":{},"note":"{}","refId":"{}"}}"#,
            txn.txn_id, txn.payer_vpa, txn.payee_vpa,
            txn.amount_paise, txn.note, txn.ref_id
        );
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",&format!("{}/collect", self.upi_url),
                   "-H","Content-Type: application/json",
                   "-d",&body,"--max-time","15"])
            .output().ok()?;
        if out.status.success() { Some(txn.txn_id) } else { None }
    }

    /// Check transaction status
    pub fn check_status(&self, txn_id: &str) -> UpiStatus {
        let out = std::process::Command::new("curl")
            .args(["-sf",&format!("{}/status/{}", self.upi_url, txn_id),"--max-time","10"])
            .output();
        match out {
            Ok(o) if o.status.success() => {
                let resp = String::from_utf8_lossy(&o.stdout);
                if resp.contains("SUCCESS") { UpiStatus::Success }
                else if resp.contains("FAILURE") || resp.contains("FAILED") { UpiStatus::Failed }
                else if resp.contains("REVERSED") { UpiStatus::Reversed }
                else { UpiStatus::Pending }
            }
            _ => UpiStatus::Pending,
        }
    }
}

// ── GST / IRN (Invoice Reference Number) ──────────────────────────────────
#[derive(Debug, Clone)]
pub struct GstInvoice {
    pub seller_gstin: String,
    pub buyer_gstin:  String,
    pub invoice_no:   String,
    pub invoice_date: String,
    pub total_value:  f64,
    pub items:        Vec<GstLineItem>,
    pub irn:          Option<String>,
    pub qr_code:      Option<String>,
}

#[derive(Debug, Clone)]
pub struct GstLineItem {
    pub hsn_code:   String,
    pub description: String,
    pub qty:        f64,
    pub rate:       f64,
    pub taxable_val: f64,
    pub cgst_rate:  f64,
    pub sgst_rate:  f64,
    pub igst_rate:  f64,
}

pub struct GstClient {
    pub gstin:       String,
    pub client_id:   String,
    pub irn_url:     String,   // NIC e-Invoice: https://api.einvoice1.gst.gov.in
    pub access_token: Option<String>,
}

impl GstClient {
    pub fn new(gstin: &str, client_id: &str) -> Self {
        Self {
            gstin:       gstin.to_owned(),
            client_id:   client_id.to_owned(),
            irn_url:     "https://api.einvoice1.gst.gov.in".to_owned(),
            access_token: None,
        }
    }

    pub fn sandbox() -> Self {
        let mut c = Self::new("29AAFCD5862R1ZR", "sigma-sandbox");
        c.irn_url = "https://einv-apisandbox.nic.in".to_owned();
        c
    }

    pub fn authenticate(&mut self, username: &str, password: &str) -> bool {
        let body = format!(r#"{{"UserName":"{}","Password":"{}","AppKey":"{}","ForceRefreshAccessToken":"Y"}}"#,
                           username, password, self.client_id);
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",&format!("{}/eivital/v1.04/Auth", self.irn_url),
                   "-H","Content-Type: application/json","-d",&body,"--max-time","10"])
            .output();
        match out {
            Ok(o) if o.status.success() => {
                let resp = String::from_utf8_lossy(&o.stdout);
                if let Some(tok) = extract_json(&resp, "AuthToken") {
                    self.access_token = Some(tok); return true;
                }
                false
            }
            _ => false,
        }
    }

    /// Generate IRN for an invoice
    pub fn generate_irn(&self, invoice: &mut GstInvoice) -> bool {
        let token = match &self.access_token {
            Some(t) => t.clone(),
            None => return false,
        };
        let items_json: Vec<String> = invoice.items.iter().map(|item| {
            format!(r#"{{"HsnCd":"{}","Desc":"{}","Qty":{},"UnitPrice":{},"TotAmt":{},"AssAmt":{},"GstRt":{},"CgstAmt":{},"SgstAmt":{},"IgstAmt":{}}}"#,
                    item.hsn_code, item.description, item.qty, item.rate,
                    item.qty * item.rate, item.taxable_val,
                    item.cgst_rate + item.sgst_rate + item.igst_rate,
                    item.taxable_val * item.cgst_rate / 100.0,
                    item.taxable_val * item.sgst_rate / 100.0,
                    item.taxable_val * item.igst_rate / 100.0)
        }).collect();
        let body = format!(
            r#"{{"Version":"1.1","TranDtls":{{"TaxSch":"GST","SupTyp":"B2B"}},"DocDtls":{{"Typ":"INV","No":"{}","Dt":"{}"}},"SellerDtls":{{"Gstin":"{}","TrdNm":"SigmaOS"}},"BuyerDtls":{{"Gstin":"{}","TrdNm":"Buyer"}},"ItemList":[{}],"ValDtls":{{"TotInvVal":{}}}}}"#,
            invoice.invoice_no, invoice.invoice_date,
            invoice.seller_gstin, invoice.buyer_gstin,
            items_json.join(","), invoice.total_value
        );
        let out = std::process::Command::new("curl")
            .args(["-sf","-X","POST",&format!("{}/eicore/v1.03/Invoice", self.irn_url),
                   "-H","Content-Type: application/json",
                   "-H",&format!("AuthToken: {}", token),
                   "-d",&body,"--max-time","15"])
            .output();
        match out {
            Ok(o) if o.status.success() => {
                let resp = String::from_utf8_lossy(&o.stdout);
                if let Some(irn) = extract_json(&resp, "Irn") {
                    invoice.irn = Some(irn);
                    invoice.qr_code = extract_json(&resp, "SignedQRCode");
                    return true;
                }
                false
            }
            _ => false,
        }
    }
}

// ── e-RUPI (one-time voucher) ─────────────────────────────────────────────
pub struct ERupiVoucher {
    pub voucher_id:  String,
    pub amount_paise: u64,
    pub beneficiary: String,
    pub purpose:     String,
    pub expiry:      String,
    pub status:      String,
}

impl ERupiVoucher {
    pub fn create(amount_paise: u64, beneficiary: &str, purpose: &str) -> Self {
        Self {
            voucher_id:   format!("erupi-sigma-{}", timestamp_millis()),
            amount_paise,
            beneficiary:  beneficiary.to_owned(),
            purpose:      purpose.to_owned(),
            expiry:       "2026-12-31".to_owned(),
            status:       "ACTIVE".to_owned(),
        }
    }
}

// ── Helpers ────────────────────────────────────────────────────────────────
fn extract_json(json: &str, key: &str) -> Option<String> {
    let search = format!("\"{}\":\"", key);
    let pos = json.find(&search)?;
    let rest = &json[pos + search.len()..];
    let end = rest.find('"')?;
    Some(rest[..end].to_owned())
}

fn urlencoded(s: &str) -> String {
    s.chars().map(|c| match c {
        'A'..='Z'|'a'..='z'|'0'..='9'|'-'|'_'|'.'|'~' => c.to_string(),
        ' ' => "+".to_owned(),
        c => format!("%{:02X}", c as u32),
    }).collect()
}

fn timestamp_millis() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

// ── CLI ────────────────────────────────────────────────────────────────────
pub fn india_stack_cmd(args: &[String]) {
    if args.is_empty() {
        eprintln!("sigma-india — India Stack Integration\n\
            Usage:\n\
            sigma-india abdm search <abha-id>\n\
            sigma-india upi pay <vpa> <amount-inr> <note>\n\
            sigma-india upi status <txn-id>\n\
            sigma-india gst irn <invoice-json>\n\
            sigma-india erupi create <amount-inr> <beneficiary> <purpose>\
        ");
        return;
    }
    match args[0].as_str() {
        "abdm" if args.len() > 2 && args[1] == "search" => {
            let mut client = AbdmClient::sandbox();
            if let Some(p) = client.search_patient(&args[2]) {
                println!("ABHA: {}\nName: {}\nDOB: {}\nGender: {}",
                         p.health_id, p.name, p.dob, p.gender);
            } else {
                println!("Patient not found or sandbox auth required.");
            }
        }
        "upi" if args.len() > 2 => {
            let client = UpiClient::new("sigma-merchant", "sigmaos@ybl");
            match args[1].as_str() {
                "pay" if args.len() > 4 => {
                    let amount_paise = (args[3].parse::<f64>().unwrap_or(0.0) * 100.0) as u64;
                    let uri = client.payment_uri(amount_paise, &args[4], &format!("ref{}", timestamp_millis()));
                    println!("UPI Payment URI:\n{}", uri);
                }
                "status" => {
                    println!("{:?}", client.check_status(&args[2]));
                }
                _ => {}
            }
        }
        "gst" => {
            let client = GstClient::sandbox();
            println!("GST client ready: {}", client.irn_url);
            println!("Authenticate with: sigma-india gst auth <username> <password>");
        }
        "erupi" if args.len() > 4 && args[1] == "create" => {
            let amount_paise = (args[2].parse::<f64>().unwrap_or(0.0) * 100.0) as u64;
            let v = ERupiVoucher::create(amount_paise, &args[3], &args[4]);
            println!("e-RUPI Voucher: {}\nAmount: ₹{:.2}\nBeneficiary: {}\nPurpose: {}",
                     v.voucher_id, amount_paise as f64/100.0, v.beneficiary, v.purpose);
        }
        _ => println!("Unknown india command. Run: sigma-india"),
    }
}

File: userland/india/sigma_india_stack.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 SigmaOS Project
#
# userland/india/sigma_india_stack.py — India Stack API Client
#
# Implements real API clients for:
#   - ABDM FHIR (Ayushman Bharat Digital Mission — health records)
#   - GST IRN (Invoice Reference Number via NIC portal)
#   - UPI Autopay mandate creation
#   - DigiLocker document fetch
#   - Bhashini NMT translation API
#   - NSDL/PAN verification
#
# Phase H implementation — requires Phase G network stack (TCP/HTTPS).
# Uses urllib.request (no external deps) for HTTP calls.

import hashlib
import json
import os
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, asdict
from typing import Optional

# ── Configuration ──────────────────────────────────────────────────────────
ABDM_BASE    = os.environ.get("ABDM_BASE",    "https://dev.abdm.gov.in/gateway")
GST_BASE     = os.environ.get("GST_BASE",     "https://einvoice1-uat.nic.in/ei/api")
UPI_BASE     = os.environ.get("UPI_BASE",     "https://api.npci.org.in/upi/autopay")
DIGILOCKER   = os.environ.get("DIGILOCKER",   "https://digilocker.meripehchaan.gov.in")
BHASHINI_BASE= os.environ.get("BHASHINI_BASE","https://dhruva-api.bhashini.gov.in/services/inference")
NSDL_BASE    = os.environ.get("NSDL_BASE",    "https://tin.tin.nsdl.com/panstatus/rest")

CLIENT_ID    = os.environ.get("SIGMA_ABDM_CLIENT_ID",    "sigma-os-dev")
CLIENT_SEC   = os.environ.get("SIGMA_ABDM_CLIENT_SECRET","changeme")
GST_USERNAME = os.environ.get("SIGMA_GST_USERNAME", "testgstin")
GST_APP_KEY  = os.environ.get("SIGMA_GST_APP_KEY",  "testappkey")

# ── HTTP helpers ──────────────────────────────────────────────────────────

def _post(url: str, body: dict, headers: dict = None, timeout: int = 10) -> dict:
    data = json.dumps(body).encode()
    h = {"Content-Type": "application/json", "Accept": "application/json"}
    if headers: h.update(headers)
    req = urllib.request.Request(url, data=data, headers=h, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as r:
            return json.loads(r.read().decode())
    except urllib.error.HTTPError as e:
        return {"error": e.code, "message": e.read().decode()[:200]}
    except Exception as e:
        return {"error": "network", "message": str(e)}

def _get(url: str, headers: dict = None, timeout: int = 10) -> dict:
    h = {"Accept": "application/json"}
    if headers: h.update(headers)
    req = urllib.request.Request(url, headers=h)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as r:
            return json.loads(r.read().decode())
    except urllib.error.HTTPError as e:
        return {"error": e.code, "message": e.read().decode()[:200]}
    except Exception as e:
        return {"error": "network", "message": str(e)}

# ── 1. ABDM — Ayushman Bharat Digital Mission ──────────────────────────────

class AbdmClient:
    """Ayushman Bharat Digital Mission — FHIR-based health record API."""

    def __init__(self):
        self._token: Optional[str] = None
        self._token_expiry: float = 0

    def _ensure_token(self) -> bool:
        if self._token and time.time() < self._token_expiry:
            return True
        resp = _post(f"{ABDM_BASE}/v0.5/sessions", {
            "clientId": CLIENT_ID,
            "clientSecret": CLIENT_SEC,
        })
        if "accessToken" in resp:
            self._token = resp["accessToken"]
            self._token_expiry = time.time() + resp.get("expiresIn", 3600) - 60
            return True
        return False

    def _auth_headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self._token}",
            "X-CM-ID": "sbx",
        }

    def generate_health_id(self, aadhaar_number: str, mobile: str) -> dict:
        """Create an ABHA (Ayushman Bharat Health Account) ID."""
        if not self._ensure_token():
            return {"error": "auth_failed"}
        return _post(f"{ABDM_BASE}/v2/registration/aadhaar/generateOtp",
            {"aadhaar": aadhaar_number, "mobile": mobile},
            self._auth_headers())

    def fetch_health_records(self, abha_id: str, from_date: str, to_date: str) -> dict:
        """Fetch FHIR health records for an ABHA ID."""
        if not self._ensure_token():
            return {"error": "auth_failed"}
        return _post(f"{ABDM_BASE}/v0.5/health-information/cm/request",
            {
                "requestId": hashlib.md5(abha_id.encode()).hexdigest(),
                "timestamp": _iso_ts(),
                "query": {
                    "healthInformationTypes": ["OPConsultation", "DiagnosticReport", "Prescription"],
                    "dateRange": {"from": from_date, "to": to_date},
                    "patientId": abha_id,
                }
            },
            self._auth_headers())

    def link_health_facility(self, abha_id: str, hip_id: str) -> dict:
        """Link a healthcare facility (HIP) to an ABHA account."""
        if not self._ensure_token():
            return {"error": "auth_failed"}
        return _post(f"{ABDM_BASE}/v0.5/links/link/add-contexts",
            {"requestId": hashlib.md5(f"{abha_id}{hip_id}".encode()).hexdigest(),
             "timestamp": _iso_ts(),
             "link": {"accessToken": "link_token", "patient": {"id": abha_id,
                      "referenceNumber": hip_id, "careContexts": []}}},
            self._auth_headers())

    def verify_otp(self, txn_id: str, otp: str) -> dict:
        if not self._ensure_token():
            return {"error": "auth_failed"}
        return _post(f"{ABDM_BASE}/v2/registration/aadhaar/verifyOTP",
            {"txnId": txn_id, "otp": otp},
            self._auth_headers())

# ── 2. GST — e-Invoice IRN API ─────────────────────────────────────────────

class GstIrnClient:
    """GST e-Invoice API — generates Invoice Reference Numbers via NIC portal."""

    def __init__(self):
        self._auth_token: Optional[str] = None
        self._sek: Optional[str] = None   # Session Encryption Key

    def authenticate(self, gstin: str) -> bool:
        """Authenticate with GST e-Invoice API."""
        resp = _post(f"{GST_BASE}/auth",
            {"Username": GST_USERNAME, "ForceRefreshAccessToken": "true",
             "Gstin": gstin, "OtpRequest": "true"})
        if resp.get("Status") == 1:
            self._auth_token = resp.get("AuthToken")
            self._sek        = resp.get("Sek")
            return True
        return False

    def generate_irn(self, invoice: dict) -> dict:
        """Generate an IRN for an invoice. invoice = B2B invoice dict."""
        if not self._auth_token:
            return {"error": "not_authenticated"}
        # The real API requires AES encryption of the payload using SEK
        # Here we format the request structure correctly
        payload = {
            "Version": "1.1",
            "TranDtls": {
                "TaxSch": "GST",
                "SupTyp": invoice.get("supply_type", "B2B"),
                "RegRev": "N",
            },
            "DocDtls": {
                "Typ": invoice.get("doc_type", "INV"),
                "No": invoice.get("invoice_no"),
                "Dt": invoice.get("invoice_date"),
            },
            "SellerDtls": invoice.get("seller", {}),
            "BuyerDtls":  invoice.get("buyer",  {}),
            "ItemList":   invoice.get("items",   []),
            "ValDtls": {
                "AssVal": invoice.get("assessable_value", 0),
                "CgstVal": invoice.get("cgst", 0),
                "SgstVal": invoice.get("sgst", 0),
                "IgstVal": invoice.get("igst", 0),
                "TotInvVal": invoice.get("total_value", 0),
            },
        }
        return _post(f"{GST_BASE}/invoice",
            {"Data": json.dumps(payload)},
            {"authtoken": self._auth_token, "gstin": invoice.get("seller_gstin", "")})

    def get_irn_details(self, irn: str) -> dict:
        """Fetch details of an existing IRN."""
        return _get(f"{GST_BASE}/invoice/irn/{irn}",
            {"authtoken": self._auth_token or ""})

    def cancel_irn(self, irn: str, reason: int, remarks: str) -> dict:
        """Cancel an IRN (within 24 hours of generation)."""
        return _post(f"{GST_BASE}/invoice/cancel",
            {"Irn": irn, "CnlRsn": str(reason), "CnlRem": remarks},
            {"authtoken": self._auth_token or ""})

# ── 3. UPI Autopay ────────────────────────────────────────────────────────

class UpiAutopayClient:
    """NPCI UPI Autopay — recurring payment mandate management."""

    def create_mandate(
        self, payer_vpa: str, payee_vpa: str,
        amount_paise: int, frequency: str,
        start_date: str, end_date: str,
        purpose: str = "Subscription",
    ) -> dict:
        """Create a recurring UPI Autopay mandate."""
        mandate_id = f"SIGMA{int(time.time()*1000)}"
        return _post(f"{UPI_BASE}/create",
            {
                "mandateId": mandate_id,
                "payerVpa":  payer_vpa,
                "payeeVpa":  payee_vpa,
                "amount":    amount_paise,
                "currency":  "INR",
                "frequency": frequency,    # DAILY, WEEKLY, MONTHLY, ASPRESENTED
                "startDate": start_date,
                "endDate":   end_date,
                "purpose":   purpose,
                "txnInitType": "CREATE",
            })

    def revoke_mandate(self, mandate_id: str, payer_vpa: str) -> dict:
        return _post(f"{UPI_BASE}/revoke",
            {"mandateId": mandate_id, "payerVpa": payer_vpa})

    def mandate_status(self, mandate_id: str) -> dict:
        return _get(f"{UPI_BASE}/status/{mandate_id}")

    def initiate_debit(self, mandate_id: str, amount_paise: int, txn_note: str) -> dict:
        """Trigger a debit against an existing mandate."""
        return _post(f"{UPI_BASE}/debit",
            {"mandateId": mandate_id, "amount": amount_paise,
             "txnNote": txn_note, "txnId": f"STXN{int(time.time()*1000)}"})

# ── 4. DigiLocker ─────────────────────────────────────────────────────────

class DigiLockerClient:
    """DigiLocker — fetch Aadhaar, PAN, driving licence, degree certificates."""

    def __init__(self, access_token: str = ""):
        self._token = access_token

    def fetch_document(self, doc_type: str, doc_id: str) -> dict:
        """Fetch a document. doc_type: ADHAR, PANCR, DRVLC, etc."""
        return _get(f"{DIGILOCKER}/1/xml/pullfile",
            {"Authorization": f"Bearer {self._token}",
             "doctype": doc_type, "docid": doc_id})

    def list_documents(self) -> dict:
        """List all documents in a user's DigiLocker."""
        return _get(f"{DIGILOCKER}/1/xml/files",
            {"Authorization": f"Bearer {self._token}"})

    def issue_certificate(self, department: str, cert_type: str, data: dict) -> dict:
        """Trigger certificate issuance from a government department."""
        return _post(f"{DIGILOCKER}/1/xml/issuedoc",
            {"dept": department, "certType": cert_type, "data": data},
            {"Authorization": f"Bearer {self._token}"})

# ── 5. Bhashini — Indian Language Translation ─────────────────────────────

class BhashiniClient:
    """Bhashini ULCA/Dhruva API — NMT translation, ASR, TTS for Indian languages."""

    LANG_CODES = {
        "hindi": "hi", "tamil": "ta", "telugu": "te", "kannada": "kn",
        "malayalam": "ml", "bengali": "bn", "marathi": "mr", "gujarati": "gu",
        "punjabi": "pa", "odia": "or", "assamese": "as", "urdu": "ur",
        "english": "en",
    }

    def __init__(self, api_key: str = ""):
        self._api_key = api_key or os.environ.get("SIGMA_BHASHINI_KEY", "")

    def translate(self, text: str, src_lang: str, tgt_lang: str) -> str:
        """Translate text from src_lang to tgt_lang."""
        src = self.LANG_CODES.get(src_lang, src_lang)
        tgt = self.LANG_CODES.get(tgt_lang, tgt_lang)
        resp = _post(f"{BHASHINI_BASE}/nmt/inference",
            {
                "pipelineTasks": [{
                    "taskType": "translation",
                    "config": {"language": {"sourceLanguage": src, "targetLanguage": tgt}},
                }],
                "inputData": {"input": [{"source": text}]},
            },
            {"Authorization": self._api_key})
        try:
            return resp["pipelineResponse"][0]["output"][0]["target"]
        except (KeyError, IndexError):
            return f"[translation unavailable: {resp.get('error', 'unknown error')}]"

    def asr_transcribe(self, audio_base64: str, lang: str) -> str:
        """Transcribe speech to text (ASR)."""
        lang_code = self.LANG_CODES.get(lang, lang)
        resp = _post(f"{BHASHINI_BASE}/asr/inference",
            {
                "pipelineTasks": [{
                    "taskType": "asr",
                    "config": {"language": {"sourceLanguage": lang_code}},
                }],
                "inputData": {"audio": [{"audioContent": audio_base64}]},
            },
            {"Authorization": self._api_key})
        try:
            return resp["pipelineResponse"][0]["output"][0]["source"]
        except (KeyError, IndexError):
            return ""

    def tts_synthesize(self, text: str, lang: str, gender: str = "female") -> bytes:
        """Synthesize speech from text (TTS). Returns WAV bytes."""
        lang_code = self.LANG_CODES.get(lang, lang)
        resp = _post(f"{BHASHINI_BASE}/tts/inference",
            {
                "pipelineTasks": [{
                    "taskType": "tts",
                    "config": {"language": {"sourceLanguage": lang_code},
                               "gender": gender},
                }],
                "inputData": {"input": [{"source": text}]},
            },
            {"Authorization": self._api_key})
        try:
            import base64
            audio_b64 = resp["pipelineResponse"][0]["audio"][0]["audioContent"]
            return base64.b64decode(audio_b64)
        except (KeyError, IndexError, Exception):
            return b""

# ── 6. NSDL PAN verification ──────────────────────────────────────────────

class NsdlPanClient:
    """NSDL PAN card status verification."""

    def verify_pan(self, pan: str) -> dict:
        """Verify a PAN card number's validity and status."""
        pan = pan.upper().strip()
        # Basic format check: AAAAA9999A
        import re
        if not re.match(r'^[A-Z]{5}[0-9]{4}[A-Z]$', pan):
            return {"valid": False, "error": "invalid_format", "pan": pan}
        return _post(f"{NSDL_BASE}/json",
            {"getpanstatus": {"pan_number_1": pan, "pan_application_type": "pan"}})

# ── Helper ─────────────────────────────────────────────────────────────────
def _iso_ts() -> str:
    import datetime
    return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.000Z")

# ── CLI demo ──────────────────────────────────────────────────────────────
def main():
    import argparse
    parser = argparse.ArgumentParser(prog="sigma-india-stack", description="India Stack API Client")
    sub = parser.add_subparsers(dest="cmd", required=True)

    sub.add_parser("abdm-auth",      help="Test ABDM authentication")
    sub.add_parser("gst-auth",       help="Test GST e-Invoice authentication")
    p_tr = sub.add_parser("translate", help="Translate text via Bhashini")
    p_tr.add_argument("text"); p_tr.add_argument("--from", default="english", dest="src")
    p_tr.add_argument("--to",   default="hindi",   dest="tgt")
    p_pan = sub.add_parser("pan",    help="Verify a PAN number")
    p_pan.add_argument("pan_number")
    sub.add_parser("upi-demo",       help="Demo UPI Autopay mandate creation")

    args = parser.parse_args()

    if args.cmd == "abdm-auth":
        client = AbdmClient()
        ok = client._ensure_token()
        print(f"ABDM auth: {'✓ success' if ok else '✗ failed (check env vars SIGMA_ABDM_CLIENT_ID/SECRET)'}")

    elif args.cmd == "gst-auth":
        client = GstIrnClient()
        ok = client.authenticate(os.environ.get("SIGMA_GSTIN", "29AAABB1111D1Z5"))
        print(f"GST auth: {'✓ success' if ok else '✗ failed (check env vars)'}")

    elif args.cmd == "translate":
        client = BhashiniClient()
        result = client.translate(args.text, args.src, args.tgt)
        print(f"  {args.src} → {args.tgt}: {result}")

    elif args.cmd == "pan":
        client = NsdlPanClient()
        result = client.verify_pan(args.pan_number)
        print(json.dumps(result, indent=2))

    elif args.cmd == "upi-demo":
        client = UpiAutopayClient()
        result = client.create_mandate(
            payer_vpa="user@oksbi", payee_vpa="merchant@hdfc",
            amount_paise=9900, frequency="MONTHLY",
            start_date="2026-08-01", end_date="2027-07-31",
            purpose="SigmaOS Subscription",
        )
        print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()
